pulsar/internal/commands.go

```diff
@@ -221,9 +222,21 @@ func serializeBatch(wb Buffer,
 	cmdSend *pb.BaseCommand,
 	msgMetadata *pb.MessageMetadata,
 	uncompressedPayload Buffer,
-	compressionProvider compression.Provider) {
+	compressionProvider compression.Provider,
+	encryptor crypto.Encryptor) {
 	// Wire format
 	// [TOTAL_SIZE] [CMD_SIZE][CMD] [MAGIC_NUMBER][CHECKSUM] [METADATA_SIZE][METADATA] [PAYLOAD]
+
+	// compress the payload
+	compressedPayload := compressionProvider.Compress(nil, uncompressedPayload.ReadableSlice())
+
+	// encrypt the compressed payload
+	encryptedPayload, err := encryptor.Encrypt(compressedPayload, crypto.NewMessageMetadataSupplier(msgMetadata))
+	if err != nil {
+		// error occurred while encrypting the payload, ProducerCryptoFailureAction is set to Fail
+		panic(fmt.Sprintf("Encryption of message failed, ProducerCryptoFailureAction is set to Fail. Error :%v", err))
```

Comment:
What do other clients do here? This could crash a server.

Reply:
Java client throws Exception [PulsarClientException](https://github.com/apache/pulsar/blob/34a04b5fa4692cb4716df61e740795467f76451b/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L679)

Reply:
It depends on end user if they are catching the exception. Since it is java, it is possible to catch the exception.

I think it is proper action to `panic` here because  `ProducerCryptoFailureAction=Fail` set by the user in config. 

I do see other panics in the same function [serializeBatch](https://github.com/apache/pulsar-client-go/blob/master/pulsar/internal/commands.go#L220) and if they are recoverable then this panic can also be recovered.

Reply:
In java and c++, only the message is marked as failed to send and will NOT explicitly crash the application.  We should follow the same semantics here.

Reply:
Thanks @jerrypeng  @cckellogg  Yes, I agree It is better to not crash the application.

But I'm just trying to understand how java client do not crash the application.

I found below steps where in it throws exception and possibly crash the application.

- [ProducerBase send](https://github.com/apache/pulsar/blob/34a04b5fa4692cb4716df61e740795467f76451b/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBase.java#L63)
- [TypedMessaggeBuilderImpl send](https://github.com/apache/pulsar/blob/34a04b5fa4692cb4716df61e740795467f76451b/pulsar-client/src/main/java/org/apache/pulsar/client/impl/TypedMessageBuilderImpl.java#L91)
- [ProducerImpl completable future](https://github.com/apache/pulsar/blob/34a04b5fa4692cb4716df61e740795467f76451b/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L466)
- [ProducerImpl completing future with exception](https://github.com/apache/pulsar/blob/34a04b5fa4692cb4716df61e740795467f76451b/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L318)
- [ProducerImpl encryptMessage](https://github.com/apache/pulsar/blob/34a04b5fa4692cb4716df61e740795467f76451b/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L679)

Reply:
And when compared  with java client , I don't see any existing way to mark message as failed in go client. 
So need suggestions  on achieving similar thing in go client.

Reply:
This code will check if the message was added and if not it returns an error through the callback.
https://github.com/apache/pulsar-client-go/blob/master/pulsar/producer_partition.go#L434

I think we should we should look into changing the BatchBuilder Add message to return an error as opposed to a bool (this could be a separate MR). (https://github.com/apache/pulsar-client-go/blob/master/pulsar/internal/batch_builder.go#L47). There will be a little refactoring for some of the functions. If this returned an error then we could fail the message and return that to the client with a meaningful message. The bool will still work it's just the message returned is not very helpful.